01-getting-started/projects/final/app/src/main/java/com/raywenderlich/podplay/repository/PodcastRepo.kt

```diff
@@ -0,0 +1,150 @@
+ *   or information technology.  Permission for such use, copying, modification,
+ *   merger, publication, distribution, sublicensing, creation of derivative works,
+ *   or sale is expressly withheld.
+ *   
+ *   This project and source code may use libraries or frameworks that are
+ *   released under various Open-Source licenses. Use of those libraries and
+ *   frameworks are governed by their own individual licenses.
+ *   
+ *   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *   THE SOFTWARE.
+ */
+
+package com.raywenderlich.podplay.repository
+
+import androidx.lifecycle.LiveData
+import com.raywenderlich.podplay.db.PodcastDao
+import com.raywenderlich.podplay.model.Episode
+import com.raywenderlich.podplay.model.Podcast
+import com.raywenderlich.podplay.service.RssFeedResponse
+import com.raywenderlich.podplay.service.RssFeedService
+import com.raywenderlich.podplay.util.DateUtils
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+
+class PodcastRepo(private var feedService: RssFeedService,
+                  private var podcastDao: PodcastDao) {
+
+  suspend fun getPodcast(feedUrl: String): Podcast? {
+    val podcastLocal = podcastDao.loadPodcast(feedUrl)
+    if (podcastLocal != null) {
+      podcastLocal.id?.let {
+        podcastLocal.episodes = podcastDao.loadEpisodes(it)
+        return podcastLocal
+      }
+    }
+    var podcast: Podcast? = null
+    val feedResponse = feedService.getFeed(feedUrl)
+    if (feedResponse != null) {
+      podcast = rssResponseToPodcast(feedUrl, "", feedResponse)
+    }
+    return podcast
+  }
+
+  private fun rssItemsToEpisodes(episodeResponses: List<RssFeedResponse.EpisodeResponse>): List<Episode> {
+    return episodeResponses.map {
+      Episode(
+          it.guid ?: "",
+          null,
+          it.title ?: "",
+          it.description ?: "",
+          it.url ?: "",
+          it.type ?: "",
+          DateUtils.xmlDateToDate(it.pubDate),
+          it.duration ?: ""
+      )
+    }
+  }
+
+  private fun rssResponseToPodcast(
+      feedUrl: String, imageUrl: String, rssResponse: RssFeedResponse
+  ): Podcast? {
+    val items = rssResponse.episodes ?: return null
+    val description = if (rssResponse.description == "")
+      rssResponse.summary else rssResponse.description
+    return Podcast(null, feedUrl, rssResponse.title, description, imageUrl,
+        rssResponse.lastUpdated, episodes = rssItemsToEpisodes(items))
+  }
+
+  fun save(podcast: Podcast) {
```

Comment:
This should be a suspend function and called inside view model scope.

We should avoid GlobalScope because it needs to be maintained manually. I know most of these code lines are from the previous authors, however, this is a big problem. I don't see these coroutines being handled anywhere. Also this is not a place for using it because this is not a global job. And I do see more of these, **please refactor.**